app/api/Consignment.scala

```diff
@@ -0,0 +1,326 @@
+      (__ \ "qualifierOfIdentificationDetails" \ "authorisationNumber").readNullable[String] and
+      (__ \ "qualifierOfIdentificationDetails" \ "additionalIdentifier").readNullable[String] and
+      (__ \ "qualifierOfIdentificationDetails" \ "unlocode" \ "unLocodeExtendedCode").readNullable[String] and
+      (__ \ "qualifierOfIdentificationDetails" \ "customsOffice").readNullable[CustomsOfficeType01](customsOfficeType01.reads) and
+      (__ \ "qualifierOfIdentificationDetails" \ "coordinates").readNullable[GNSSType](gnssType.reads) and
+      (__ \ "qualifierOfIdentificationDetails" \ "identificationNumber").readNullable[EconomicOperatorType03](economicOperatorType03.reads) and
+      (__ \ "qualifierOfIdentificationDetails").read[Option[AddressType14]](addressType14.reads) and
+      (__ \ "qualifierOfIdentificationDetails").read[Option[PostcodeAddressType02]](postcodeAddressType02.reads) and
+      (__ \ "contactPerson").readNullable[ContactPersonType06](contactPersonType06.reads)
+  )(LocationOfGoodsType01.apply _)
+
+}
+
+object economicOperatorType03 {
+
+  implicit val reads: Reads[EconomicOperatorType03] =
+    __.read[String].map(EconomicOperatorType03)
+}
+
+object customsOfficeType01 {
+
+  implicit val reads: Reads[CustomsOfficeType01] =
+    (__ \ "id").read[String].map(CustomsOfficeType01)
+}
+
+object gnssType {
+
+  implicit val reads: Reads[GNSSType] = (
+    (__ \ "latitude").read[String] and
+      (__ \ "longitude").read[String]
+  )(GNSSType.apply _)
+}
+
+object addressType14 {
+
+  implicit val reads: Reads[Option[AddressType14]] = (
+    (__ \ "address" \ "numberAndStreet").readNullable[String] and
+      (__ \ "address" \ "postalCode").readNullable[String] and
+      (__ \ "address" \ "city").readNullable[String] and
+      (__ \ "country" \ "code").readNullable[String]
+  ).tupled.map {
+    case (Some(streetAndNumber), postcode, Some(city), Some(country)) =>
+      Some(AddressType14(streetAndNumber, postcode, city, country))
+    case _ => None
+  }
+
+}
+
+object addressType01 {
+
+  implicit val reads: Reads[Option[AddressType01]] = (
+    (__ \ "address" \ "numberAndStreet").readNullable[String] and
+      (__ \ "address" \ "postalCode").readNullable[String] and
+      (__ \ "address" \ "city").readNullable[String]
+  ).tupled.map {
+    case (Some(streetAndNumber), postcode, Some(city)) =>
+      Some(AddressType01(streetAndNumber, postcode, city))
+    case _ => None
+  }
+
+}
+
+object postcodeAddressType02 {
+
+  implicit val reads: Reads[Option[PostcodeAddressType02]] = (
+    (__ \ "postalCode" \ "streetNumber").readNullable[String] and
+      (__ \ "postalCode" \ "postalCode").readNullable[String] and
+      (__ \ "postalCode" \ "country" \ "code").readNullable[String]
+  ).tupled.map {
+    case (streetAndNumber, Some(postcode), Some(country)) =>
+      Some(PostcodeAddressType02(streetAndNumber, postcode, country))
+    case _ => None
+  }
```

Comment:
I think this could be refactored to
```
implicit val reads: Reads[PostcodeAddressType02] = (
    (__ \ "streetNumber").readNullable[String] and
      (__ \ "postalCode").readNullable[String] and
      (__ \ "country" \ "code").readNullable[String]
  )(PostcodeAddressType02.apply _)
```

Then in the locationOfGoodsType01 reads we'd have
`(__ \ "qualifierOfIdentificationDetails" \ "postalCode").readNullable[PostcodeAddressType02](postcodeAddressType02.reads) and`